File: tests/geo.test.js
/**
 * Tests for the distance arithmetic behind a geofenced punch.
 *
 * Ported from `manna_hr/tests/test_rules.py`, case for case, when the Python
 * app was removed on 31 August 2026.
 *
 *     npm test
 */

import test from "node:test";
import assert from "node:assert/strict";

import * as geo from "../client/src/lib/geo.js";

test("zero, zero is not a place", () => {
	// (0, 0) is in the Atlantic and is what an unset Float field reads as. A
	// punch measured against it would be half a world from every gate.
	assert.equal(geo.isRealCoordinate(0, 0), false);
	assert.equal(geo.isRealCoordinate(null, null), false);
	assert.equal(geo.isRealCoordinate(undefined, undefined), false);
	assert.equal(geo.isRealCoordinate(91, 0), false);
	assert.equal(geo.isRealCoordinate("not a number", 0), false);
	assert.equal(geo.isRealCoordinate(9.9312, 76.2673), true);
});

test("distance between two known points", () => {
	// Kochi to Thrissur, about 70 km by air.
	const metres = geo.metresBetween(9.9312, 76.2673, 10.5276, 76.2144);
	assert.ok(metres > 65_000 && metres < 72_000, `${metres} m is not about 70 km`);
});

test("a point is zero metres from itself", () => {
	assert.equal(geo.metresBetween(9.9312, 76.2673, 9.9312, 76.2673), 0);
});

test("antipodal points do not produce NaN", () => {
	// Rounding can push the haversine term a hair above 1, and Math.asin of
	// that is NaN — which propagates silently instead of raising. Clamped.
	const metres = geo.metresBetween(0, 0, 0, 180);
	assert.ok(Number.isFinite(metres) && metres > 0, `got ${metres}`);
});

test("the bounding box never falls inside the circle", () => {
	// The box only narrows the list; the haversine then decides. A box a little
	// too big costs extra comparisons, a box too small silently drops a match —
	// so the only acceptable error is outwards.
	const lat = 9.9312;
	const lng = 76.2673;
	const radius = 300.0;
	const { minLat, maxLat, minLng, maxLng } = geo.boundingBox(lat, lng, radius);

	assert.ok(geo.metresBetween(lat, lng, maxLat, lng) >= radius);
	assert.ok(geo.metresBetween(lat, lng, minLat, lng) >= radius);
	assert.ok(geo.metresBetween(lat, lng, lat, maxLng) >= radius);
	assert.ok(geo.metresBetween(lat, lng, lat, minLng) >= radius);
});

test("nearest skips places with no coordinate", () => {
	const uncaptured = { latitude: 0, longitude: 0 };
	const real = { latitude: 9.9312, longitude: 76.2673 };

	const found = geo.nearest(9.931, 76.267, [uncaptured, real]);
	assert.notEqual(found, null);
	assert.equal(found.place, real);
});

test("nearest returns null when nothing can be measured", () => {
	// null means "cannot tell", not "too far". A caller that treats it as a
	// refusal strands every employee whose gate was never captured.
	assert.equal(geo.nearest(9.93, 76.26, []), null);
	assert.equal(geo.nearest(9.93, 76.26, [{ latitude: 0, longitude: 0 }]), null);
});

test("distance reads the way a person would say it", () => {
	assert.equal(geo.formatDistance(240.4), "240 m");
	assert.equal(geo.formatDistance(2400), "2.4 km");
});

File: app/test/geo_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:manna_hr_app/core/geo.dart';

/// The distance arithmetic, against `manna_hr/geo.py` — three copies of one
/// formula are only worth having if they agree, and the Python side has no test
/// of its own yet, so this is currently the only place the numbers are pinned.
void main() {
  group('distance', () {
    test('the same point is no distance at all', () {
      expect(metresBetween(10.0, 76.5, 10.0, 76.5), closeTo(0, 0.001));
    });

    test('a tenth of a degree of latitude is about eleven kilometres', () {
      // 11119 m, not the 11054 the box constant would suggest: `metresBetween`
      // is a great circle on a 6371 km sphere, while `METRES_PER_DEGREE` is
      // deliberately understated so a bounding box errs outwards. The two
      // numbers are meant to differ, and this is where that is written down.
      final d = metresBetween(10.0, 76.5, 10.1, 76.5);
      expect(d, closeTo(11119.5, 1));
    });

    test('it is symmetric, whichever end you measure from', () {
      final there = metresBetween(10.0, 76.5, 10.05, 76.55);
      final back = metresBetween(10.05, 76.55, 10.0, 76.5);
      expect(there, closeTo(back, 0.001));
    });

    test('antipodal points do not come back as NaN', () {
      // Rounding can push the haversine term a hair above 1, and asin of that
      // is NaN — which then propagates silently through every comparison
      // downstream instead of raising.
      final d = metresBetween(0, 0, 0, 180);
      expect(d.isNaN, isFalse);
      expect(d, closeTo(20015086, 1000));
    });
  });

  group('whether a coordinate is a place', () {
    test('null island is not a location, it is an unset field', () {
      // (0, 0) is in the Atlantic and is what an unset Float reads as. A record
      // that was never captured must not be measured against the Gulf of Guinea.
      expect(isRealCoordinate(0, 0), isFalse);
    });

    test('a missing coordinate is not a location', () {
      expect(isRealCoordinate(null, 76.5), isFalse);
      expect(isRealCoordinate(10.0, null), isFalse);
    });

    test('an impossible coordinate is not a location', () {
      expect(isRealCoordinate(91.0, 76.5), isFalse);
      expect(isRealCoordinate(10.0, 181.0), isFalse);
    });

    test('a real one is', () {
      expect(isRealCoordinate(10.0523, 76.5312), isTrue);
    });
  });

  group('how far, in words', () {
    test('metres up close and kilometres beyond', () {
      expect(formatDistance(42.4), '42 m');
      expect(formatDistance(999), '999 m');
      expect(formatDistance(1400), '1.4 km');
    });
  });

  test('a bounding box is rounded outwards, never inwards', () {
    // A box a little too big costs a few extra comparisons; a box a little too
    // small silently drops a match before anything ever measures it.
    expect(latSpanForMetres(1000) * 110540, greaterThan(1000));
  });
}
